### PR TITLE
flake: add overlay and packages to build the home-manager package

### PR DESCRIPTION
**Pull request by Kloenk**
### Description
add overlay and package atribute to the flake.nix, so the home-manager binary can be build via flakes

### Checklist

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/CONTRIBUTING.md#commits) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

CC: @kisik21

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -3,23 +3,39 @@
...
+
+      forAllSystems = f: nixpkgs.lib.genAttrs systems (system: f system);
+
+      # Memoize nixpkgs for different platforms for efficiency.
+      nixpkgsFor = forAllSystems (system:
+        import nixpkgs {
+          inherit system;
+          overlays = [ self.overlay ];
+        });
+    in rec {
```
Comment (Kloenk) [inline flake.nix:18]: I belive `rec` should be removed here, and self should be used instead where it is needed

**Comment by kisik21** [comment]
The binary isn't working with flakes tho

**Comment by Kloenk** [comment]
On my system both are not working?? I will try on an non nixos system, where I use home-manager

**Comment by Kloenk** [comment]
> On my system both are not working?? I will try on an non nixos system, where I use home-manager

I can't get flakes to work at all under that system at the moment

**Comment by kisik21** [comment]
@Kloenk that's totally weird, look at my flake: https://gitlab.com/vikanezrimaya/nix-flake - it should work! it does work! it works!

**Comment by Kloenk** [comment]
> @Kloenk that's totally weird, look at my flake: https://gitlab.com/vikanezrimaya/nix-flake - it should work! it does work! it works! and due to flakes being totally reproducible, it should work on your system too

The non nixos system just died of faulty kernel, so it was nothing related to nix/flakes.

And why it does not work on my nixos, I have no idea. I only use the nixos module for managing my users, and never invoked `home-manager` manually.

**Comment by Kloenk** [comment]
invoked some more home-manager on my nixos.
`home-manager generations` is working, but `home-manager build` and `home-manager news` is not working.

It gives me the error:

> kloenk@barahir ~/p/home-manager> result/bin/home-manager build
error: value is a function while a set was expected, at /nix/store/h3vi495yys7khirhwp93cphiy3c144ab-source/home-manager/home-manager.nix:36:19
(use '--show-trace' to show detailed location information)
error: value is a function while a set was expected, at /nix/store/h3vi495yys7khirhwp93cphiy3c144ab-source/home-manager/home-manager.nix:85:4
result/bin/home-manager: line 137: /tmp/home-manager-build.sH44jSwkXu/news-info.sh: No such file or directory

**Comment by kisik21** [comment]
Perfectly normal. Flakes are supposed to be pure functions, and home-manager was relying on impurities at some places, e.g. `builtins.getEnv` and stuff. While porting home-manager to flakes, @bqv purified some of these impurities by turning them into pure functions. When porting the home-manager configuration derivation as a library function, I used function arguments to additionally purify more of these impurities.

Now we'll need to purify the binary too and make it flake-aware - and when I was porting home-manager-configuration, that wasn't my objective since I use my own way of deploying home-manager generations, and this project's `bqv-flakes` branch and Nix's flake support are still in "preview" phase.

I don't know what to do with news; since news are implemented as a soup of values and functions and I'm not really sure if they're pure and can be used with flakes.

**Comment by Kloenk** [comment]
I got the `news` and `build` command to work. Not sure how, but it seems like it is working.

**Comment by Kloenk** [comment]
Also fixed the test environment, but a test is failing, I don't see what is failing. so cannot fix it

**Comment by bqv** [comment]
In the interests of making it easy to rebase in the future, are there any commits you can squash here?

File: flake.nix (hunk 1 above)
Comment (bqv) [reply flake.nix:18]: I don't think there even is any self-reference anywhere. Can probably be removed

**Comment by Kloenk** [comment]
I manually rebased this onto the gcroot feature (#1170) And this makes the tests succeding:

[travis](https://travis-ci.com/Kloenk/home-manager)
[master](https://github.com/kloenk/home-manager/tree/master)

**Comment by Kloenk** [comment]
rebased it to the current branch

File: flake.nix (hunk 1 above)
Comment (Kloenk) [reply flake.nix]: removed it with the current rebase

**Comment by Kloenk** [comment]
I also rewrote the test framework (at the moment only wrote a test for git) to be available on hydra.

@rycee see if you like this framework and want to change to it.

**Comment by Kloenk** [comment]
I updated all other tests, so this would close #1172

**Comment by rycee** [comment]
Sorry, haven't really followed this thread. Just some loose throughts:

- As far as I know the news derivation should be pure.

- The `builtins.getEnv` problem should be fixed even not considering flakes and I plan to remove their use starting at state version 20.09. Options like `home.homeDirectory` will then be undefined by default and must be entered by the user in their configuration.

- I like parts of the updated test framework, in particular that it works with Hydra but I'm absolutely against pulling in the framework into this repo. The test framework should stay in the nmt repo since I'm not keen to maintain two testing frameworks 🙂

  As far as I can tell, it is the `toplevel` attributes that makes it work with Hydra? I.e. to have a derivation that fails to build if the test fails. That seems easy to add to nmt.

**Comment by Kloenk** [comment]
> * I like parts of the updated test framework, in particular that it works with Hydra but I'm absolutely against pulling in the framework into this repo. The test framework should stay in the nmt repo since I'm not keen to maintain two testing frameworks slightly_smiling_face
>   As far as I can tell, it is the `toplevel` attributes that makes it work with Hydra? I.e. to have a derivation that fails to build if the test fails. That seems easy to add to nmt.

Yes, the test fails if the derivation is failing to build. I will try to understand `nmt` so I can patch it. Unfortionaly I don't have a Gitlab account at the moment, so I only can mail you the patch, and not make a pr. If you are faster, I could also give your version a try in my hydra, as long as it has a flake.nix to import it into hydra.

Furthermore I would advise to pull nmt via the flake.nix.

**Comment by Kloenk** [comment]
Applied my changes under https://github.com/Kloenk/nmt 

Or here are the Files

[nmt.zip](https://github.com/rycee/home-manager/files/4672545/nmt.zip)

**Comment by Kloenk** [comment]
I reinstalled my hydra, so the current build are under https://hydra.kloenk.de/jobset/home-manager/flake-overlay.

The new hydra also has a miniio backend, so home-manager and the tests are downloadable from https://cache.kloenk.de with the sig: `cache.kloenk.de:ea1cL0mwRMABkALTC/cYV84V0eoL1UWkj3e2TvS4Y6o=`

**Comment by terlar** [comment]
Seems like quite a bit of work went in to this, is there something we can salvage?

**Comment by Kloenk** [comment]
I think this was dependent on mnt, but that never got merged. So not sure

**Pull request closed without merging**